cpp/leaky_core.cc

```diff
@@ -181,10 +184,31 @@ bool InputAvailable(int fd) {
   NotImplemented();
 }
 
-void SignalState_AfterForkingChild() {
-  signal(SIGQUIT, SIG_DFL);
-  signal(SIGPIPE, SIG_DFL);
-  signal(SIGTSTP, SIG_DFL);
+static void signal_handler(int sig_num) {
+  assert(gSignalHandlers != nullptr);
+  SignalHandler* handler = gSignalHandlers->get(sig_num);
+  assert(handler != nullptr);
+  handler->Run(sig_num);
```

Comment:
This part isn't safe, as you alluded to in the comment

The handler can contain a `command_t` that will be run by the interpreter right?

It shouldn't be run directly from the signal handler, for several reasons, like it's not safe to malloc() from a signal handler

It should set a flag, and then the main loop will run it

(This is how CPython, bash, dash, etc. all work)

Reply:
> This part isn't safe, as you alluded to in the comment
> 
> The handler can contain a `command_t` that will be run by the interpreter right?

Ah no this isn't my intention here. Right now the only meaningful thing that implements `SignalHandler` is `TrapState` whose `Run()` simply appens the `command_t` for the trap to the run list for the main loop to run later